Update node version dependencies

hackerrdave opened the pull request:
-  Unit tests fail when use Node v8 (due to request mismatched headers), bumping Nock library to use major version 9 fixes this issue
 - Nock version 9 is incompatible with Node versions < 4 due to use of `Object.assign`, this PR provides a polyfill for that function when running the tests and that function is not otherwise defined.

griffinmyers commented:
Hey @hackerrdave, thanks for submitting!

Looks like we're quite a bit out of date with Node support.  While we're at this, I'd like to actually get this repo caught up to speed with current support ([ref](https://github.com/nodejs/Release)). 

That will mean: 

* Dropping support for versions `< v6.x`
* Adding support for `v8.x` and `v10.x`
* Removing API to support non-native Promises + supporting documentation
* Not needing an `Object.assign` polyfill.

There's enough going on here I'd prefer to submit a patch for this today (I'll have to cut a release for you anyways).  If you're _really_ eager though, let me know and feel free to pursue those additional changes--beware I can be a picky reviewer 😉.

hackerrdave commented:
ah yea sounds good! I can take a look at the remaining items later today

griffinmyers commented:
Sounds good--just had a look over the README--if we're dropping support for non-native Promises, we should probably also drop support for node-style callbacks (one async paradigm is better than two).  Turning into quite the yak-shave (lots of README to update and such..) feel free to dump back over to me anytime.

hackerrdave commented:
@griffinmyers updated the MR to include:
 

- Removing Callback Support
- Removing Custom non-native Promise support
- Support for Node < v6 (no more polyfill, etc.)
- Support for Node 10 (updating Buffer api calls)
- Updated all tests & documentation to reflect API updates

Updated package.json as well to reflect v3.0.0 for the module, should I also update the CHANGELOG?

Thanks
Dave

griffinmyers commented:
Hey @hackerrdave -- sorry I thought I left a comment on this yesterday.  Thanks for the incredibly thorough patch!   I'm very excited to take a look either later today or tomorrow!

griffinmyers commented:
It looks like some test are failing on node `v8.9.1` and `v10.7.1`.  

Would you mind opening up this PR against a branch of this repo rather than as a fork?  I _think_ that will get TravisCI to kick in so we can see tests against supported versions run.

hackerrdave commented:
@griffinmyers I enabled Travis on my forked repo (wasn't going to be able to push to this repo to open a PR) and now I can see the tests run against the 3 versions of Node (6, 8, 10)

griffinmyers commented:
Ah! Awesome, thanks @hackerrdave.  Taking a look now.

hackerrdave commented on package.json:
```diff
@@ -15,15 +15,14 @@
-    "bluebird": "^3.4.1",
-    "eslint": "^2.13.1",
-    "expect.js": "^0.3.1",
+    "chai": "^4.0.0",
```
ah yea def makes sense! I swapped out `expect.js` with `chai` solely b/c `expect.js` hasn't been updated in a few years - I'll verify it works with Node 10, and if so reinstate it

hackerrdave commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@button/button-client-node",
-  "version": "2.8.0",
+  "version": "3.0.0",
```
ah gotcha - yea will do!

hackerrdave commented on README.md:
```diff
@@ -136,11 +90,11 @@ We currently expose the following resources to manage:
-    // ...
-});
+client.accounts.all()
+  .then((accounts) => handleAccounts(accounts))
```
ah yea - I meant to go back and do that, will do

hackerrdave commented on lib/request.js:
```diff
@@ -12,118 +11,106 @@ function request(timeout, secure) {
+  return function _request(options, data) {
+    return new Promise((resolve, reject) => {
+      let req;
+      try {
```
sounds good - will do

hackerrdave commented on lib/request.js:
```diff
@@ -12,118 +11,106 @@ function request(timeout, secure) {
-    req.on('response', responseHandler(callback));
-    req.on('error', errorHandler(callback));
+        req.on('response', responseHandler(resolve, reject));
+        req.on('error', errorHandler(reject));
```
ah yea good point!

hackerrdave commented on test/lib/request-test.js:
```diff
@@ -33,410 +34,361 @@ describe('lib/#request', function() {
-      expect(res).to.eql(null);
-      scope.done();
-      done();
+    }).catch((err) => {
```
ah yep! that's a good call

hackerrdave commented:
@griffinmyers think I addressed all your comments - let me know if I missed anything! One thing to note is that 1 test fails (`handles timeout errors`) on Node v10.7, but the same test passes on Node v10.6 (why I updated the travisci settings to use specifically 10.6). The `timeoutHandler` function never fires on 10.7 - possibly a bug in latest version of Node?

griffinmyers commented:
Ah, just seeing this @hackerrdave (think my github notifications need some tuning)! ACK--will take a look tomorrow.

hackerrdave commented on CHANGELOG.md:
```diff
@@ -1,40 +1,54 @@
 2.8.0 December 5, 2017
-  - Add Links API
+
+- Add Links API
```
oh weird not sure (possibly an errant auto-change by my editor) - I'll fix

griffinmyers commented:
Hey @hackerrdave -- I poked into the node 10 issue a bit today, and ended up submitting an upstream patch to nock [here](https://github.com/nock/nock/pull/1182).  Let's give that a few days and see if the maintainers are eager to move forward--otherwise we can consider some hacks here just to get tests passing.

griffinmyers commented:
Woohoo, our patch to nock was merged!  Can you please apply this patch?

```diff
diff --git a/package.json b/package.json
index 1a19b17..383e671 100644
--- a/package.json
+++ b/package.json
@@ -19,7 +19,7 @@
     "expect.js": "^0.3.1",
     "istanbul": "^0.4.4",
     "mocha": "^5.0.0",
-    "nock": "^9.0.0"
+    "nock": "https://github.com/nock/nock.git#80da59cab336352834baeda8c5653e63249d3dbc"
   },
   "engines": {
     "node": ">=6.0"
diff --git a/test/lib/request-test.js b/test/lib/request-test.js
index 9357c6d..1abf412 100644
--- a/test/lib/request-test.js
+++ b/test/lib/request-test.js
@@ -162,7 +162,7 @@ describe('lib/#request', function() {
 
     this.scope = nock('https://' + hostname + ':443')
       .get(path)
-      .delayConnection(100)
+      .socketDelay(10000)
       .reply(200, successResponse({}));
 
     return timeoutRequest({
```

hackerrdave commented:
great thanks!!
